src/Common/src/Interop/Windows/SspiCli/SafeDeleteContext.cs

```diff
@@ -33,6 +34,7 @@ protected SafeDeleteContext() : base(IntPtr.Zero, true)
 
         public override bool IsInvalid
         {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Why is this needed?

Reply:
Doesn't inline

Reply:
> Doesn't inline

... and?  Is that a 0.00000001% improvement, or does it actually move a needle on anything beyond a microbenchmark?

The JIT has heuristics for determining when to inline.  If there's a general pattern that it misses or scores "incorrectly" then we should solve it in the JIT and leave AggressiveInlining for those cases where a) it looks like, but is provably different, from a thing the JIT says no to and b) it actually moves the needle on real code.

Reply:
> ... and? Is that a 0.00000001% improvement

Not inlining pushes IsAuthenticated back to being slower than the crypto for small messages; due to the high ratio of its call count and bools having messy asm

![image](https://user-images.githubusercontent.com/1142958/62962840-ac7dad00-bdf7-11e9-9443-ffeff0f30e8f.png)


>  does it actually move a needle on anything beyond a microbenchmark?

I'm not microbenchmarking; this is from our real application which is entire raison d'être is very high speed websocket messages.

>  If there's a general pattern that it misses or scores "incorrectly" then we should solve it in the JIT 

I'd prefer the JIT's handling of bools to be improved as when inlined it could produce far better asm than it currently does. Currently it does a test, then materilizes that test to a register to pass it back as the return; which it then does a test on to get the bool back, which the vestiges of which still remain after inlining which should be optimzed away https://github.com/dotnet/corefx/pull/40265#issuecomment-520901302

> a) it looks like, but is provably different, from a thing the JIT says no to

And this is what the traces are showing; maybe it doesn't like bools for the reason above. The common path is to execute almost all of the tests as the common state is valid and since anything can make it invalid it needs to check everything so they aren't conditional calls in the common path.

> b) it actually moves the needle on real code.

This is real code and every trace in this PR has been from real code; and while WebSockets and Kestrel's streams are bothering me, SslStream is my main nemesis:

![image](https://user-images.githubusercontent.com/1142958/62962626-2c574780-bdf7-11e9-8800-3861ddef8ae8.png)

And this is one of many steps to address its issues

Reply:
And, just to be sure, with the AggressiveInline on IsAuthenticated it actually removes 250ms from this test (presumably the overhead of calling a function 96,000 times) (and plays accounting games with the other 110ms)?

'Cuz that I'd admit is moving the needle :wink:

Reply:
Well its a 4 function chain so is  96,000 x 4 = 384,000 calls; or 12 additional method calls per read and write; vs 4 with the inlined version.

Call counting may be skewing the results; since the JIT doesn't optimize the inlined bools very well without the ternary https://github.com/dotnet/corefx/pull/40265#discussion_r313679833; however each of the method is fairly minimal and it does give the JIT the opportunity to optimize the bools better.